# Introducción a Pandas

Pandas es una librería construida encima de NumPy. Nos proporciona una eficiente implementación de objetos de datos construidos a base del arreglo de NumPy, además de otras operaciones para manipular datos. Este tipo de operaciones se conocen como **data wrangling** - los pasos requeridos para preparar la data a fin de que pueda ser consumida para extraer información y construir modelos.

La preparación de data es lo que toma más tiempo en un proyecto de ciencia de datos.

Existen dos (2) componentes fundamentales en Pandas: los objetos de **Series** y **DataFrame**. Una **Series** es esencialmente una columna y un **DataFrame** es una table multi-dimensional constituida a base de una colección de Series. La misma puede consistir de tipos de datos heterogéneos y también puede contener data faltante.

Los objetos de Pandas pueden ser pensados como versiones ampliadas de los arreglos de NumPy, donde las filas y columnas son identificadas con **labels**, e.g. "edad", en vez de índices, e.g. "1".

## Series desde Listas y Arreglos

### El objeto Series

Un objeto Series en Pandas es un arreglo unidimensional (1D) de data indexada 

---

esencialmente una columna.  Puede ser creada desde una lista o un arreglo utilizando el método `pd.Series()`.

In [ ]:
import pandas as pd 
series = pd.Series([0, 1, 2, 3])
print(series)

0    0
1    1
2    2
3    3
dtype: int64


Notamos que el objeto Series consiste en: 
* secuencia de valores
* secuencia de índices

Los valores son simplemente un arreglo NumPy, pero el índice es un objeto parecido a un arreglo  de tipo `pd.Index`. Los valores pueden ser accedidos con el índice correspondiente, como veremos a continuación.

In [ ]:
import pandas as pd 
series = pd.Series([0, 1, 2, 3, 4, 5])

print("Valores:", series.values)
print("Indices:", series.index, "\n")

print(series[1], "\n")   # Un solo valor

print(series[1:4]) # Un rango de valores

Valores: [0 1 2 3 4 5]
Indices: RangeIndex(start=0, stop=6, step=1) 

1 

1    1
2    2
3    3
dtype: int64


**Pregunta:** ¿Cómo puede acceder a los valores [2,3,4] del arreglo ([0, 1, 2, 3, 4, 5]), utilizando el índice correspondiente?

In [ ]:
import pandas as pd
series = pd.Series([0, 1, 2, 3, 4, 5])

print("Valores:", series.values)

print(series[2:5]) 

No obstante, el índice no tiene que ser un número entero, puede consistir en valores de cualquier tipo, i.e. podemos utilizar cadenas de caracteres como índices.

In [ ]:
import pandas as pd 
data = pd.Series([12, 24, 13, 54], 
                index=['a', 'b', 'c', 'd'])

print(data, "\n")
print("Valor en el indice b:", data['b'])

a    12
b    24
c    13
d    54
dtype: int64 

Valor en el indice b: 24


## Series desde Diccionarios

Podemos observar que el objeto Series se parecen a diccionarios en Python. Un diccionario es una estructura que hace un enlace entre una llave o clave y un valor; pero la Series es una estructura que hace un enlace entre llaves tipificadas y valores tipificados. Este tipo de información las hace más eficientes comparadas a un diccionario estándar. 

In [ ]:
import pandas as pd 

fruits_dict = { 'manzanas': 10,
                'narajanjas': 8,
                'bananas': 3,
                'fresas': 20
               }

fruits = pd.Series(fruits_dict)
print("Valor para manazanas: ", fruits['manzanas'], "\n")

# Series también permiten operacioens de arreglos como rebanar o slicing:
print(fruits['bananas':'fresas'])

Valor para manazanas:  10 

bananas     3
fresas     20
dtype: int64


### El objeto DataFrame

El objeto *Series* es esencialmente una columna, sin embargo, el objeto *DataFrame* es un a table multi-dimensional construida a base de una colección de objetos Series. *DataFrame* nos permite guardar y manipular data tabular donde las filas consisten en observaciones y las columnas representan variables.

Existen varias formas de crear un objeto DatFrame utilizando `pd.DataFrame()`. Por ejemplo, podemos crear un *DataFrame* de las siguientes formas:
* pasando múltiples Series al objeto *DataFrame*
* convirtiendo un diccionario a un *DataFrame*
* importando data desde un archivo *csv*



#### Construyendo un DataFrame desde una Series

Podemos crear un *DataFrame* desde una *Series* al pasarle el objeto Series como dato de entrada el método de creación de un *DataFrame*, en adición a otros datos de entrada opcional como el parámetro `column` que nos permite nombrar columnas.

In [ ]:
import pandas as pd 

data_s1 = pd.Series(
    [12, 24, 33, 15], 
    index=['manzanas', 'bananas', 'fresas', 'naranjas'])

# 'quantity' is the name for our column
dataframe1 = pd.DataFrame(data_s1, columns=['quantity'])
print(dataframe1)

          quantity
manzanas        12
bananas         24
fresas          33
naranjas        15


#### Construyendo un DataFrame desde un Diccionario

Podemos construir un *DataFrame* desde una lista de diccionarios. Digamos que tenemos un diccionario con países, sus capitales y alguna otra variable como población, tamaño del país, número de escuelas, etc.)

In [ ]:
dict = {"país": ["Noruega", "Suecia", "España", "Francia"],
       "capital": ["Oslo", "Estocolmo", "Madrid", "Paris"],
       "OtraColumna": ["100", "200", "300", "400"]}

data = pd.DataFrame(dict)
print(data)

      país    capital OtraColumna
0  Noruega       Oslo         100
1   Suecia  Estocolmo         200
2   España     Madrid         300
3  Francia      Paris         400


#### Construyendo un DataFrame al importar Data desde un archivo

Es sencillo cargar datos a un *DataFrame* desde varios archivos con distintos formatos, e.g. csv, excel, json. Vamos a estar importando datos para analizar un dataset específico. 

In [ ]:
import pandas as pd 

# Si tenemos un archivo llamado data1.csv en nuestro directorio
df = pd.read_csv('data1.csv')

# Si tenemos un archivo en formato json
df = pd.read_json('data2.json')

**Pregunta:** ¿Cúal es la diferencia entre un DataFrame construido desde una Series y un DataFrame construido desde un Diccionario?

## Explorando un Dataset

Vamos a trabajar con el dataset de películas denominado IMDB. Este dataset esta públicamente disponible y contiene información de 14,762 películas. Cada fila consiste en una película, y cada película tiene información como título, año de publicación, director, número de premios, puntaje, duración, etc.  

### Leyendo data desde un CSV

El dataset de IMDB lo podemos encontrar [aquí](https://www.kaggle.com/PromptCloudHQ/imdb-data/data).

Nota: Si el archivo de datos se encuentra en Google Drive, tenemos que conectarlo con Google Colab, importando `drive` de `google.colab` y ejecutar el método `drive.mount("/content/drive")`.Accede al enlace proporcionado y copy/paste el token de acceso.

Si todo marcha bien deberias ver el siguiente mensaje:
```
Mounted at /content/drive
```

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly&response_type=code

Enter your authorization code:
··········
Mounted at /content/drive


In [ ]:
# revisemos que se ha montado el Google Drive
!ls

drive  sample_data


Podemos cargarlo utilizando el método de creación de *DataFrame* insertando la ruta del archivo como dato de entrada del método `pd.read_csv('ruta/del/archivo').

La manera más fácil de obtener la ruta del archivo en tu Drive es accediendo al *sidebar* de la izquierda, a fin de encontrar tu archivo dentro del directorio `/content/drive/My Drive`.

Si necesitas una ayuda visual, accede a este artículo [aquí](https://medium.com/ml-book/simplest-way-to-open-files-from-google-drive-in-google-colab-fae14810674).

In [ ]:
import pandas as pd

ruta = '/content/drive/My Drive/Colab Notebooks/data/IMDB-Movie-Data.csv'

# leemos los datos en el archivo csv
peliculas_df = pd.read_csv(ruta)

Hemos creado un objeto DataFrame, `peliculas_df`, con índices pre-configurados. No obstante, ¿qué sucede si queremos acceder a cada fila por título de la película y no por número de índice? Podemos hacerlo si configuramos el título como nuestro índice. Esto lo podemos hacer de dos maneras: 
1. pasando el nombre de la columna como parámetro adicional a `index_col` en el método `read_csv`, cuando cargamos el archivo
2. después de haber cargado el archivo, podemos ejecutar el método `set_index()`

El índice nos permite hacer búsquedas rápidas y eficientes, toda vez que buscando filas por valores de índice es como buscar valores en un diccionario a base de una llave o clave.

In [ ]:
# 1. cuando cargamos el archivo
peliculas_df_indice_x_titulo = pd.read_csv(ruta, index_col='Title')

# 2. despues de haber cargado el archivo
peliculas_df_indice_x_titulo = peliculas_df.set_index('Title')

### Viendo la Data

**Después** de haber creado el objeto DataFrame, el primer paso es darle una ojeada a la data, a fin de crear una imagen mental de la misma y llegar a familiarizarnos con ella.

Podemos utilizar el método `head()` para visualizar las primeras filas de nuestro dataset. Esto retorna las primeras cinco (5) filas del DataFrame por preconfiguración, pero podemos pasarle el número de filas que deseamos como parámetro de entrada.

In [ ]:
# las primeras 5 filas
peliculas_df.head()

Rank                    Title  ... Revenue (Millions) Metascore
0     1  Guardians of the Galaxy  ...             333.13      76.0
1     2               Prometheus  ...             126.46      65.0
2     3                    Split  ...             138.12      62.0
3     4                     Sing  ...             270.32      59.0
4     5            Suicide Squad  ...             325.02      40.0

[5 rows x 12 columns]

In [ ]:
# las primeras 10 filas
peliculas_df.head(10)

Rank                    Title  ... Revenue (Millions) Metascore
0     1  Guardians of the Galaxy  ...             333.13      76.0
1     2               Prometheus  ...             126.46      65.0
2     3                    Split  ...             138.12      62.0
3     4                     Sing  ...             270.32      59.0
4     5            Suicide Squad  ...             325.02      40.0
5     6           The Great Wall  ...              45.13      42.0
6     7               La La Land  ...             151.06      93.0
7     8                 Mindhorn  ...                NaN      71.0
8     9       The Lost City of Z  ...               8.01      78.0
9    10               Passengers  ...             100.01      41.0

[10 rows x 12 columns]

También podemos ver las últimas filas del dataset con el método `tail()`, que funciona de la misma manera que `head()`

In [ ]:
# las ultimas 5 filas
peliculas_df.tail()

Rank                   Title  ... Revenue (Millions) Metascore
995   996    Secret in Their Eyes  ...                NaN      45.0
996   997         Hostel: Part II  ...              17.54      46.0
997   998  Step Up 2: The Streets  ...              58.01      50.0
998   999            Search Party  ...                NaN      22.0
999  1000              Nine Lives  ...              19.64      11.0

[5 rows x 12 columns]

**Pregunta:** ¿Cómo se pueden retornar las últimas tres (3) filas del DataFrame?

In [ ]:
# las ultimas 3 filas
peliculas_df.tail(3)

Rank                   Title  ... Revenue (Millions) Metascore
997   998  Step Up 2: The Streets  ...              58.01      50.0
998   999            Search Party  ...                NaN      22.0
999  1000              Nine Lives  ...              19.64      11.0

[3 rows x 12 columns]

**Pregunta:** ¿Qué método permite visualizar las primeras filas de un dataset?


### Consiguiendo información de la Data

Como primer paso, es recomendado conseguir un buen pantallazo de la data. Vamos a utilizar 2 métodos sumamente útiles para esto: 
- `info()` -- nos permite ver detalles esenciales como el número de filas y columnas, el número de índices, el tipo de dato de cada columna, el número de valores que no son nulos, y la memoria utilizada por el DataFrame.
- `describe()` -- nos proporciona un breve análisis estadístico de nuestra data

In [ ]:
peliculas_df_indice_x_titulo.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, Guardians of the Galaxy to Nine Lives
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Genre               1000 non-null   object 
 2   Description         1000 non-null   object 
 3   Director            1000 non-null   object 
 4   Actors              1000 non-null   object 
 5   Year                1000 non-null   int64  
 6   Runtime (Minutes)   1000 non-null   int64  
 7   Rating              1000 non-null   float64
 8   Votes               1000 non-null   int64  
 9   Revenue (Millions)  872 non-null    float64
 10  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(4)
memory usage: 133.8+ KB


Importante destacar que `info()` nos permite ver que:
- el dataset consiste de 1000 filas y 11 columnas
- utiliza 93KB de memoria
- las columnas *Revenue* y *Metascore* contiene valores nulos

El manejo de valores nulos es muy importante en la fase de preparación de datos y cualquier proyecto de ciencia de datos, toda vez que los algoritmos de aprendizaje de maquina y métodos de análisis de datos no pueden manipular valores nulos por si mismos.

El `Dtype` es útil para aclarar cualquier duda en cuanto el tipo de dato de cada columna. A veces una columna que esperamos sea de `numero entero` son en realidad `cadenas de caracteres` y en este debemos convertirlo a `numero` antes de realizar la operación de análisis.

In [ ]:
# para saber el numero de filas y columnas
peliculas_df_indice_x_titulo.shape

(1000, 11)

**Pregunta:** ¿Qué proporciona el método describe()?

In [ ]:
peliculas_df_indice_x_titulo.describe()

Rank         Year  ...  Revenue (Millions)   Metascore
count  1000.000000  1000.000000  ...          872.000000  936.000000
mean    500.500000  2012.783000  ...           82.956376   58.985043
std     288.819436     3.205962  ...          103.253540   17.194757
min       1.000000  2006.000000  ...            0.000000   11.000000
25%     250.750000  2010.000000  ...           13.270000   47.000000
50%     500.500000  2014.000000  ...           47.985000   59.500000
75%     750.250000  2016.000000  ...          113.715000   72.000000
max    1000.000000  2016.000000  ...          936.630000  100.000000

[8 rows x 7 columns]

El método `describe()` nos ha proporcionado bastante información de alto nivel de la data. Por ejemplo, podemos ver que:
- el data set consiste de películas entre 2006 (min) y 2016 (max)
- el ingreso más alto generado por una película es de 936.63 millones de dólares americanos, mientras que el más bajo fue de 82.9 millones de dólares americanos.

## Operaciones de DataFrame - Seleccionar, Rebanar y Filtrar

#### Trabajando con Columnas

In [ ]:
# extraemos una columna utilizando su nombre
genero_df = peliculas_df['Genre']  # retorna un objeto Series

Lo anterior retorna un objeto Series. Si queremos obtener un objeto DataFrame, debemos pasarla el nombre de la columna como una lista, ver a continuación.

In [ ]:
# We can select any column using its label:

# To obtain a Series as output
col_como_series = peliculas_df['Genre']

# Imprime el tipo de objeto y las primeras 5 filas
print(type(col_como_series))
col_como_series.head()


# Obtenemos un objeto DataFrame
col_como_df = peliculas_df[['Genre']]

# Imprime el tipo de objeto y las primeras 5 filas
print(type(col_como_df))
col_como_df.head()

<class 'pandas.core.series.Series'>
<class 'pandas.core.frame.DataFrame'>


Genre
0   Action,Adventure,Sci-Fi
1  Adventure,Mystery,Sci-Fi
2           Horror,Thriller
3   Animation,Comedy,Family
4  Action,Adventure,Fantasy

**Pregunta:** ¿Cómo se extraen muchas columnas del DataFrame?

In [ ]:
# podemos extraer muchas columnas
muchas_columnas = peliculas_df_indice_x_titulo[['Genre', 'Rating', 'Revenue (Millions)']]

muchas_columnas.head()

Genre  Rating  Revenue (Millions)
Title                                                                        
Guardians of the Galaxy   Action,Adventure,Sci-Fi     8.1              333.13
Prometheus               Adventure,Mystery,Sci-Fi     7.0              126.46
Split                             Horror,Thriller     7.3              138.12
Sing                      Animation,Comedy,Family     7.2              270.32
Suicide Squad            Action,Adventure,Fantasy     6.2              325.02

Observemos también la diferencia entre utilizar un DataFrame indexado por título y el otro indexado por números. En este último ejemplo, el título de las películas se muestra en vez de los números de fila.

#### Trabajando con Filas

Podemos trabajar con fila con los siguientes indexadores:
- `loc` -- nos permite indexar y rebanar haciendo referencia explícita al índice pero ubicándolo por nombre. Por ejemplo, en el DataFrame indexado por título, podemos utilizar el título para seleccionar la fila requerida.
- `iloc` -- nos permite indexar y rebanar haciendo referencia implícita a la indexación nativa de Python pero ubicandolo por el número del índice. En el caso de nuestro DataFrame, le pasamos el número de índice de la película que deseamos seleccionar.


In [ ]:
# Siendo explícito con el índice, "Guardians of the Galaxy":
gog = peliculas_df_indice_x_titulo.loc["Guardians of the Galaxy"]

# dando el índice numerico de "Guardians of the Galaxy":
gog = peliculas_df_indice_x_titulo.iloc[0]

gog

Rank                                                                  1
Genre                                           Action,Adventure,Sci-Fi
Description           A group of intergalactic criminals are forced ...
Director                                                     James Gunn
Actors                Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...
Year                                                               2014
Runtime (Minutes)                                                   121
Rating                                                              8.1
Votes                                                            757074
Revenue (Millions)                                               333.13
Metascore                                                            76
Name: Guardians of the Galaxy, dtype: object

In [ ]:
# podemos tambien rebanar con múltiples filas
muchas_filas = peliculas_df_indice_x_titulo['Guardians of the Galaxy': 'Sing']
muchas_filas = peliculas_df_indice_x_titulo[0:4]

muchas_filas

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0

[4 rows x 11 columns]

También podemos seleccionar filas y columnas. El primer índice hace referencia a filas, mientras que el segundo a columnas.

In [ ]:
# selecciona todas las filas hasta Sing y todas las columnas hasta Director
peliculas_df_indice_x_titulo.loc[:'Sing', :'Director']
peliculas_df_indice_x_titulo.iloc[:4, :3]

Rank  ...                                        Description
Title                          ...                                                   
Guardians of the Galaxy     1  ...  A group of intergalactic criminals are forced ...
Prometheus                  2  ...  Following clues to the origin of mankind, a te...
Split                       3  ...  Three girls are kidnapped by a man with a diag...
Sing                        4  ...  In a city of humanoid animals, a hustling thea...

[4 rows x 3 columns]

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0
...                       ...  ...       ...
Secret in Their Eyes      996  ...      45.0
Hostel: Part II           997  ...      46.0
Step Up 2: The Streets    998  ...      50.0
Search Party              999  ...      22.0
Nine Lives               1000  ...      11.0

[1000 rows x 11 columns]

**Pregunta:** ¿Cuál es la diferencia entre el método loc y el método iloc?

### Seleccionar condicionalmente y filtrando

Hemos seleccionado filas y columnas con índices específicos. Sin embargo, ¿cómo hacemos para seleccionar cuando no conocemos el índice de antemano o cuando queremos filtrar a base de alguna condición? 

Por ejemplo, queremos que el DataFrame nos muestre las peliíulas desde 2016 o todas las películas que obtuvieron un rating de más de 8.0? Podemos aplicar condiciones booleanas a las columnas en el DataFrame.

In [ ]:
# todas las peliculas del 2016
peliculas_df_indice_x_titulo[peliculas_df_indice_x_titulo['Year'] == 2016]

# peliculas con un puntaje mayor a 8.0
peliculas_df_indice_x_titulo[peliculas_df_indice_x_titulo['Rating'] > 8.0 ]

Rank  ... Metascore
Title                                               ...          
Guardians of the Galaxy                          1  ...      76.0
La La Land                                       7  ...      93.0
Hacksaw Ridge                                   17  ...      71.0
Lion                                            19  ...      69.0
Bahubali: The Beginning                         27  ...       NaN
Interstellar                                    37  ...      74.0
Star Wars: Episode VII - The Force Awakens      51  ...      81.0
The Dark Knight                                 55  ...      82.0
The Prestige                                    65  ...      66.0
Mad Max: Fury Road                              68  ...      90.0
Zootopia                                        75  ...      78.0
The Avengers                                    77  ...      69.0
Inglourious Basterds                            78  ...      69.0
Inception                                       81  ...      74.0
The Wolf of Wall Street                         83  ...      75.0
Gone Girl                                       84  ...      79.0
Prisoners                                       91  ...      74.0
The Help                                        93  ...      62.0
Kimi no na wa                                   97  ...      79.0
The Departed                                   100  ...      85.0
12 Years a Slave                               112  ...      96.0
Harry Potter and the Deathly Hallows: Part 2   115  ...      87.0
Dangal                                         118  ...       NaN
The Dark Knight Rises                          125  ...      78.0
Whiplash                                       134  ...      88.0
No Country for Old Men                         137  ...      91.0
Shutter Island                                 139  ...      63.0
Room                                           144  ...      86.0
Django Unchained                               145  ...      81.0
Ah-ga-ssi                                      146  ...      84.0
Twin Peaks: The Missing Pieces                 155  ...       NaN
Spotlight                                      193  ...      93.0
Warrior                                        195  ...      71.0
Into the Wild                                  198  ...      73.0
The Imitation Game                             199  ...      73.0
Pan's Labyrinth                                231  ...      98.0
The Grand Budapest Hotel                       239  ...      88.0
Inside Out                                     242  ...      94.0
The Intouchables                               250  ...      57.0
There Will Be Blood                            300  ...      92.0
Rush                                           335  ...      75.0
The Bourne Ultimatum                           428  ...      85.0
3 Idiots                                       431  ...      67.0
Jagten                                         456  ...      76.0
The Lives of Others                            477  ...      89.0
Paint It Black                                 479  ...      71.0
Up                                             500  ...      88.0
Mommy                                          609  ...      74.0
WALL·E                                         635  ...       NaN
Relatos salvajes                               642  ...      77.0
Gran Torino                                    646  ...       NaN
Toy Story 3                                    689  ...      92.0
Hachi: A Dog's Tale                            696  ...      61.0
Incendies                                      714  ...      80.0
El secreto de sus ojos                         743  ...      80.0
PK                                             766  ...      51.0
How to Train Your Dragon                       773  ...      74.0
Koe no katachi                                 862  ...      80.0
Taare Zameen Par                               992  ...      42.0

[59 rows x 11 colum

Si queremos filtrar de manera más compleja lo podemos hacer con operadores `|` y `&`. Por ejemplo, si queremos seleccionar las ultimas películas (entre 2010-2016) que tuvieron un puntaje bajo (menos de 6.0) pero que están entre las películas que mas generaron ingresos (encima del 75%) 

In [ ]:
peliculas_df_indice_x_titulo[
    ((peliculas_df_indice_x_titulo['Year'] >= 2010) & (peliculas_df_indice_x_titulo['Year'] <= 2016))
    & (peliculas_df_indice_x_titulo['Rating'] < 6.0)
    & (peliculas_df_indice_x_titulo['Revenue (Millions)'] > peliculas_df_indice_x_titulo['Revenue (Millions)'].quantile(0.75))
]

Rank  ... Metascore
Title                                            ...          
Fifty Shades of Grey                         64  ...      46.0
Ghostbusters                                 80  ...      60.0
Transformers: Age of Extinction             127  ...      32.0
The Twilight Saga: Breaking Dawn - Part 2   367  ...      52.0
Grown Ups 2                                 395  ...      19.0
Clash of the Titans                         576  ...      39.0
Kickboxer: Vengeance                        581  ...      37.0
Teenage Mutant Ninja Turtles                658  ...      31.0
Green Lantern                               674  ...      39.0
G.I. Joe: Retaliation                       879  ...      41.0
The Twilight Saga: Breaking Dawn - Part 1   926  ...      45.0
The Twilight Saga: Eclipse                  942  ...      58.0

[12 rows x 11 columns]

El resultado nos indica que "Fifty Shades of Grey" es la película con el peor puntaje, pero con más altos ingresos. En total, hay 12 películas que dan con esas condiciones. 

Notemos también el 75% del `describe()` reportaba 113.715 millones de dólares americanos y todas las películas generaron más que eso.

## Operaciones de DataFrame - Agrupar y Ordenar

### Agrupando

Las cosas comienzan a ponerse interesante cuando agrupamos filas que cumplen cierto criterio y luego agregamos su data. 

Digamos que queremos agrupar las películas por director y ver cuanto ingreso generó cada director en total, además del puntaje promedio por director. 

Esto lo podemos hacer con la operación `groupby` en la columna de interés, seguida por el agregado correspondiente (suma o promedio)

In [ ]:
# agrupemos por director y agreguemos por suma
peliculas_df.groupby('Director')[['Revenue (Millions)']].sum()

Revenue (Millions)
Director                               
Aamir Khan                         1.20
Abdellatif Kechiche                2.20
Adam Leon                          0.00
Adam McKay                       438.14
Adam Shankman                    157.33
...                                 ...
Xavier Dolan                       3.49
Yimou Zhang                       45.13
Yorgos Lanthimos                   8.81
Zack Snyder                      975.74
Zackary Adler                      6.53

[644 rows x 1 columns]

In [ ]:
# agrupemos por director y agreguemos por promedio
peliculas_df.groupby('Director')[['Rating']].mean()

Rating
Director                   
Aamir Khan             8.50
Abdellatif Kechiche    7.80
Adam Leon              6.50
Adam McKay             7.00
Adam Shankman          6.30
...                     ...
Xavier Dolan           7.55
Yimou Zhang            6.10
Yorgos Lanthimos       7.20
Zack Snyder            7.04
Zackary Adler          5.10

[644 rows x 1 columns]

Podemos notar que Aamir Khan obtuvo el puntaje más alto, pero a la vez sus ingresos son más bajos en comparación a los otros directores.

Además de `sum()` y `mean()` Pandas también tiene otras funciones de agregación como `min(0` y `max()`.

### Ordenando

Observemos que los resultados anteriores fueron presentados en orden alfabético. Sin embargo, sería útil presentarlos por orden de ingresos generados. 

Para lograrlo, podemos utilizar el método `sort_values(columna, ascending=False)` 

In [ ]:
peliculas_df.groupby('Director')[['Revenue (Millions)']].sum().sort_values(['Revenue (Millions)'], ascending=False)

Revenue (Millions)
Director                             
J.J. Abrams                   1683.45
David Yates                   1630.51
Christopher Nolan             1515.09
Michael Bay                   1421.32
Francis Lawrence              1299.81
...                               ...
Jalil Lespert                    0.00
Jamal Hill                       0.00
James Franco                     0.00
James Lapine                     0.00
Greg Kwedar                      0.00

[644 rows x 1 columns]

In [ ]:
# si queremos obtener mayor ingresos y mayor puntaje
data_ordenada = peliculas_df_indice_x_titulo.sort_values(['Revenue (Millions)', 'Rating'], ascending=False)
data_ordenada.head(5)

Rank  ... Metascore
Title                                             ...          
Star Wars: Episode VII - The Force Awakens    51  ...      81.0
Avatar                                        88  ...      83.0
Jurassic World                                86  ...      59.0
The Avengers                                  77  ...      69.0
The Dark Knight                               55  ...      82.0

[5 rows x 11 columns]

- Director que generó más ingresos: J.J. Abrams
- Película con mejor puntaje e ingresos: Star Wars

## Operaciones de DataFrame - Tratando valores nulos y duplicados

### Tratando valores nulos

La diferencia entre data falsa y data real es que la data real casi nunca es limpia y homogénea. Un detalle en particular que debemos abordar es la presencia de valores faltantes o nulos. 

Existen dos formas de cómo se pueden presentar valores faltantes o nulos:
- None - es un tipo de objeto de Python utilizado para tratar data que hace falta en Python. Se puede utilizar en arreglos con el tipo de dato "objeto", e.g. arreglos de objetos de Python
- NaN (Not a Number) - un valor de punto flotante especial utilizado para representar data faltante. Este tipo de dato implica que podemos realizar operaciones matemáticas. Sin embargo, el resultado de una operación aritmética con NaN es otro NaN.

In [ ]:
import numpy as np
import pandas as pd

# Ejemplo con None
None_example = np.array([0, None, 2, 3])
print("dtype =", None_example.dtype)
print(None_example)

# Ejemplo con NaN
NaN_example = np.array([0, np.nan, 2, 3])
print("dtype =", NaN_example.dtype)
print(NaN_example)

# Operacion matematica con NaN retorna NaNs
print("Operaciones Aritmeticas")
print("Suma con NaNs:", NaN_example.sum())
print("Suma con None:", None_example.sum())

dtype = object
[0 None 2 3]
dtype = float64
[ 0. nan  2.  3.]
Operaciones Aritmeticas
Suma con NaNs: nan


TypeError: ignored

Afortunadamente Pandas fue construido para el manejo de `NaN` y `None` y trata los dos valores de la misma manera, clasificándolos como un valor faltante o nulo. Además, Pandas proporciona métodos para detectar, remover y reemplazar valores nulos: `isnull()`, `notnull()`, `dropna()`, y `fillna()`.

#### Detectando valores nulos

`inull()` y `notnull()` son útiles para detectar valores nulos en Pandas. Ambos retornan Booleanos encima de la data.

In [ ]:
peliculas_df_indice_x_titulo.isnull()  # alguans columnas retornan True

Rank  Genre  ...  Revenue (Millions)  Metascore
Title                                  ...                               
Guardians of the Galaxy  False  False  ...               False      False
Prometheus               False  False  ...               False      False
Split                    False  False  ...               False      False
Sing                     False  False  ...               False      False
Suicide Squad            False  False  ...               False      False
...                        ...    ...  ...                 ...        ...
Secret in Their Eyes     False  False  ...                True      False
Hostel: Part II          False  False  ...               False      False
Step Up 2: The Streets   False  False  ...               False      False
Search Party             False  False  ...                True      False
Nine Lives               False  False  ...               False      False

[1000 rows x 11 columns]

In [ ]:
# podemos contar la cantidad de valores nulos
peliculas_df_indice_x_titulo.isnull().sum()  # no sabemos los ingresos de 128 peliculos

Rank                    0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
dtype: int64

#### Removiendo valores nulos

Es fácil remover valores nulos; sin embargo, no es siempre la mejor manera de tratar valores nulos. Tenemos la opción de remover o reemplazar. En general, deberíamos remover cuando tenemos una cantidad pequeña de valores nulos, toda vez que no podemos simplemente remover los valores sin remover toda la fila o toda la columna.

El método `dropna()` nos permite remover filas o columnas. Si lo hacemos por fila o por columna depende del dataset. Por preconfiguración, este método remueve todas las filas que contiene un valore nulo y retorna un nuevo DataFrame sin alterar el original. Si queremos alterar el original entonces modficamos el parámetro `inplace=True`. Alternativamente, podemos remover todas las columnas que contienen valores nulos al especificar el parámetro `axis=1`.



In [ ]:
# remueve todas las filas con valores nulos
peliculas_df_indice_x_titulo.dropna()

Rank  ... Metascore
Title                           ...          
Guardians of the Galaxy      1  ...      76.0
Prometheus                   2  ...      65.0
Split                        3  ...      62.0
Sing                         4  ...      59.0
Suicide Squad                5  ...      40.0
...                        ...  ...       ...
Resident Evil: Afterlife   994  ...      37.0
Project X                  995  ...      48.0
Hostel: Part II            997  ...      46.0
Step Up 2: The Streets     998  ...      50.0
Nine Lives                1000  ...      11.0

[838 rows x 11 columns]

In [ ]:
# remueve todas las columnas con valores nulos
peliculas_df_indice_x_titulo.dropna(axis=1)

Rank                     Genre  ... Rating   Votes
Title                                                    ...               
Guardians of the Galaxy     1   Action,Adventure,Sci-Fi  ...    8.1  757074
Prometheus                  2  Adventure,Mystery,Sci-Fi  ...    7.0  485820
Split                       3           Horror,Thriller  ...    7.3  157606
Sing                        4   Animation,Comedy,Family  ...    7.2   60545
Suicide Squad               5  Action,Adventure,Fantasy  ...    6.2  393727
...                       ...                       ...  ...    ...     ...
Secret in Their Eyes      996       Crime,Drama,Mystery  ...    6.2   27585
Hostel: Part II           997                    Horror  ...    5.5   73152
Step Up 2: The Streets    998       Drama,Music,Romance  ...    6.2   70699
Search Party              999          Adventure,Comedy  ...    5.6    4881
Nine Lives               1000     Comedy,Family,Fantasy  ...    5.3   12435

[1000 rows x 9 columns]

Sin embargo, **¿qué significa remover los valores nulos en nuestro dataset especifico?**
- Remover 128 filas donde el ingreso es nulo y 64 filas donde el metascore es nulo. Esto implica una perdida muy grande de data, ya que hay buena data en las otras columnas de esas filas
- Al remover las columnas se remueven las columnas de ingreso y metascore por entero, lo cual no es inteligente tampoco.

Para evitar perder toda esta data, podemos escoger remover filas o columnas a base de un límite de parámetro, en donde solo se remueve si la mayoría de la data es nula. Esto se puede lograr especificando los parámetros de `how` y `thresh`, que nos permite controlar el número de nulos que se permiten en el DataFrame

In [ ]:
# Remueve las columnas donde todos los valores hacen falta
peliculas_df_indice_x_titulo.dropna(axis='columns', how='all')

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0
...                       ...  ...       ...
Secret in Their Eyes      996  ...      45.0
Hostel: Part II           997  ...      46.0
Step Up 2: The Streets    998  ...      50.0
Search Party              999  ...      22.0
Nine Lives               1000  ...      11.0

[1000 rows x 11 columns]

In [ ]:
# thresh para especificar el numero minimo de valores nulos
# para la fila/columna que se quedan
peliculas_df_indice_x_titulo.dropna(axis='rows', thresh=10)

Rank  ... Metascore
Title                          ...          
Guardians of the Galaxy     1  ...      76.0
Prometheus                  2  ...      65.0
Split                       3  ...      62.0
Sing                        4  ...      59.0
Suicide Squad               5  ...      40.0
...                       ...  ...       ...
Secret in Their Eyes      996  ...      45.0
Hostel: Part II           997  ...      46.0
Step Up 2: The Streets    998  ...      50.0
Search Party              999  ...      22.0
Nine Lives               1000  ...      11.0

[970 rows x 11 columns]

#### Reemplazando valores nulos

En vez de remover la data, podemos reemplazar los valores con un valor valido. Este nuevo valor puede ser un número o algún tipo de interpolación de algún valor bueno, como el promedio de la columna. Para eso, Pandas nos proporciona el método `fillna()`.

Por ejemplo, vamos a reemplazar los valores faltantes de la columna de ingresos con el promedio de ingresos.

In [ ]:
# consigue el valor promerdio de la columna
ingreso = peliculas_df_indice_x_titulo['Revenue (Millions)']
ingreso_promedio = ingreso.mean()

print("Promedio de ingresos:", ingreso_promedio)

# reemplazemos todos los valores faltantes con este promedio
ingreso.fillna(ingreso_promedio, inplace=True)

# consigaos el valor actualizado de nulos
peliculas_df_indice_x_titulo.isnull().sum()

Promedio de ingresos: 82.9563761467888


Rank                   0
Genre                  0
Description            0
Director               0
Actors                 0
Year                   0
Runtime (Minutes)      0
Rating                 0
Votes                  0
Revenue (Millions)     0
Metascore             64
dtype: int64

#### Manejando duplicados



Afortunadamente no tenemos data duplicada en nuestro dataset, pero ese no siempre es el caso. Para los casos menos fortuitos, Pandas tiene un método `drop_duplicates()`. Este método retorna una copia del DataFrame con los duplicados removidos, salvo que especifiquemos `inplace=True`, en cuyo caso los remueve del dataset original.

#### Creando nuevas columnas desde columnas existentes

A veces tenemos que crear nuevas columnas desde otras que ya existen. Es fácil hacer esto en Pandas.

Por ejemplo, si queremos introducir una nueva columna que tiene los Ingresos por Minuto de cada película, podemos dividir el ingreso por la cantidad de minutos y creamos una nueva columna con este valor.

In [ ]:
# We can use 'Revenue (Millions)' and 'Runtime (Minutes)' to calculate Revenue per Min for each movie:
peliculas_df_indice_x_titulo['Revenue per Min'] = \
peliculas_df_indice_x_titulo['Revenue (Millions)']/peliculas_df_indice_x_titulo['Runtime (Minutes)']

peliculas_df_indice_x_titulo.head(3)

Rank  ... Revenue per Min
Title                          ...                
Guardians of the Galaxy     1  ...        2.753140
Prometheus                  2  ...        1.019839
Split                       3  ...        1.180513

[3 rows x 12 columns]

## Operaciones de DataFrame - Tablas Pivote y Funciones



### Tablas Pivote

Hemos visto como las agrupaciones nos permite explorar relaciones dentro de un dataset. Una table pivote es una operación similar. 

La tabla pivote toma una columna de data como entrada y agrupa todas las entradas en una tabla de dos dimensiones, a fin de dar un resumen multidimensional de la data.

Por ejemplo, si queremos comparar el dinero ganado por varios directores por año. Podemos crear una table pivote utilizando `pivot_table`; podemos fijar el `index='Director'` (fila de la tabla pivote) y obtenemos la información sobre valor de ingresos fijando `columns=Year'`;

In [ ]:
# calculemos el promedio de ingresos por director
# pero utilizemos una tabla pivote
peliculas_df_indice_x_titulo.pivot_table('Revenue (Millions)', 
    index='Director',
    aggfunc='sum', 
    columns='Year'
  ).head()

Year                   2006    2007    2008  2009  ...  2013  2014   2015       2016
Director                                           ...                              
Aamir Khan              NaN    1.20     NaN   NaN  ...   NaN   NaN    NaN        NaN
Abdellatif Kechiche     NaN     NaN     NaN   NaN  ...   2.2   NaN    NaN        NaN
Adam Leon               NaN     NaN     NaN   NaN  ...   NaN   NaN    NaN  82.956376
Adam McKay           148.21     NaN  100.47   NaN  ...   NaN   NaN  70.24        NaN
Adam Shankman           NaN  118.82     NaN   NaN  ...   NaN   NaN    NaN        NaN

[5 rows x 11 columns]

El parámetro `aggfunc` controla el tipo de agregación que es aplicada (promedio por preconfiguración). Asá como en `1groupby`, esto puede ser una cadena de caracteres que representa una las muchas opciones comunes como `sum`, `mean`, `count`, `min`, `max`.

Nuestra tabla pivote nos permite observar que Aamir Khan nada más tiene ingresos en 2007. Esto puede implicar dos cosas: 2007 fue el ánico año en este período que algunas de sus películas fueron publicadas, o que simplemente no tenemos toda la data para este director. También podemos observar que Adam McKay tiene el mayor número de películas en este periodo de 10 años, con sus ingresos más altos en 2006 y los más bajos en 2015.

Con esta simple tabla pivote, podemos observar la tendencia anual de ingresos generados por director. Por tanto, son una herramienta sumamente útil en análisis de datos.

### Aplicando Funciones

Podemos aplicar funciones a un dataset con `apply()`. Este método retorna un valor después de procesar cada fila/columna de un dataset por otro método definido de forma arbitraria. 

Por ejemplo, podemos utilizar una función para clasificar películas en cuatro categorías a base de su puntaje: buenísima, buena, promedio, mala. Eso lo hacemos en dos pasos:
- define una función que determina la categoría en base al puntaje
- aplica la función al DataFrame

In [ ]:
# paso 1
def rating_bucket(x):
    if x >= 8.0:
        return "buenisima"
    elif x >= 7.0:
        return "buena"
    elif x >= 6.0:
        return "promedio"
    else:
        return "mala"

# paso 2
peliculas_df_indice_x_titulo["Puntaje_Categoria"] = movies_df_title_indexed["Rating"].apply(rating_bucket)

# paso 3 - veamos el resultado
peliculas_df_indice_x_titulo.head(10)[['Rating','Puntaje_Categoria']]

Rating Puntaje_Categoria
Title                                            
Guardians of the Galaxy     8.1         buenisima
Prometheus                  7.0             buena
Split                       7.3             buena
Sing                        7.2             buena
Suicide Squad               6.2          promedio
The Great Wall              6.1          promedio
La La Land                  8.3         buenisima
Mindhorn                    6.4          promedio
The Lost City of Z          7.1             buena
Passengers                  7.0             buena

## Otras lecturas

1. [Funcionalidades básicas ](http://pandas.pydata.org/pandas-docs/stable/getting_started/basics.html)
2. [Tutoriales](http://pandas.pydata.org/pandas-docs/stable/getting_started/tutorials.html)
3. [Libro de recetas](http://pandas.pydata.org/pandas-docs/stable/user_guide/cookbook.html#cookbook)